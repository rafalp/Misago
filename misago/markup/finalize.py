import re

from django.utils.translation import pgettext

QUOTE_HEADER_RE = re.compile(
    r"""
<div class="quote-heading" data-noquote="1">(?P<title>.*?)</div>
""".strip(),
    re.IGNORECASE | re.MULTILINE | re.DOTALL,
)

SPOILER_REVEAL_BTN = '<button class="spoiler-reveal" type="button"></button>'


def finalize_markup(post):
    post = QUOTE_HEADER_RE.sub(replace_quote_headers, post)
    post = replace_spoiler_reveal_buttons(post)
    return post


def replace_quote_headers(matchobj):
    title = matchobj.group("title")
    if title:
        quote_title = pgettext("quote title", "%(title)s has written:") % {
            "title": title
        }
    else:
        quote_title = pgettext("quote title", "Quoted message:")
    return '<div class="quote-heading" data-noquote="1">%s</div>' % quote_title


def replace_spoiler_reveal_buttons(post):
    final_btn = SPOILER_REVEAL_BTN.replace(
        "></", ">%s</" % pgettext("spoiler", "Reveal spoiler")
    )
    return post.replace(SPOILER_REVEAL_BTN, final_btn)
