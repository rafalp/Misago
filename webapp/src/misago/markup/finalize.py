import re

from django.utils.translation import gettext as _

QUOTE_HEADER_RE = re.compile(
    r"""
<div class="quote-heading">(?P<title>.*?)</div>
""".strip(),
    re.IGNORECASE | re.MULTILINE | re.DOTALL,
)

SPOILER_REVEAL_BTN = '<button class="spoiler-reveal" type="button"></button>'


def finalize_markup(post):
    post = QUOTE_HEADER_RE.sub(replace_quote_headers, post)
    post = replace_spoiler_reveal_buttons(post)
    return post


def replace_quote_headers(matchobj):
    title = matchobj.group("title")
    if title:
        quote_title = _("%(title)s has written:") % {"title": title}
    else:
        quote_title = _("Quoted message:")
    return '<div class="quote-heading">%s</div>' % quote_title


def replace_spoiler_headers(matchobj):
    title = matchobj.group("title")
    if title:
        spoiler_title = _("%(title)s:") % {"title": title}
    else:
        spoiler_title = _("Spoiler:")
    return '<div class="spoiler-heading">%s</div>' % spoiler_title


def replace_spoiler_reveal_buttons(post):
    final_btn = SPOILER_REVEAL_BTN.replace("></", ">%s</" % _("Reveal spoiler"))
    return post.replace(SPOILER_REVEAL_BTN, final_btn)
