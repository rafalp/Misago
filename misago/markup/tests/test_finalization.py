from ..finalize import finalize_markup


def test_finalization_sets_translation_strings_in_empty_quotes_headings(snapshot):
    test_text = '<div class="quote-heading" data-noquote="1">Lorem ipsum</div>'
    finalized_text = finalize_markup(test_text)
    snapshot.assert_match(finalized_text)


def test_finalization_sets_translation_strings_in_empty_quotes_headings(snapshot):
    test_text = '<div class="quote-heading" data-noquote="1"></div>'
    finalized_text = finalize_markup(test_text)
    snapshot.assert_match(finalized_text)


def test_finalization_sets_translation_strings_in_spoilers_buttons(snapshot):
    test_text = '<button class="spoiler-reveal" type="button"></button>'
    finalized_text = finalize_markup(test_text)
    snapshot.assert_match(finalized_text)
