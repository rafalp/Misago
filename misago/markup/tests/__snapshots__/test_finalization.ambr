# serializer version: 1
# name: test_finalization_sets_translation_strings_in_empty_quotes_headings
  '<div class="quote-heading" data-noquote="1">Quoted message:</div>'
# ---
# name: test_finalization_sets_translation_strings_in_spoilers_buttons
  '<button class="spoiler-reveal" type="button">Reveal spoiler</button>'
# ---
